File: 1-Data Preprocessing/Python/Feature_Scaling/feature_scaling.py
# Data Preprocessing

# Importing the libraries
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 3].values

# -------  Taking care of missing data ------------

from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer = imputer.fit(X[:, 1:3])
X[:, 1:3] = imputer.transform(X[:, 1:3])

# ---------- Encoding categorical data -------------

from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Encoding the Dependent Variable
labelencoder_Y = LabelEncoder()
Y = labelencoder_Y.fit_transform(Y)

# Encoding the Independent Variable :: We can not catogorize as 0,1,2
# as it will show comparison between countries.

labelencoder_X = LabelEncoder()
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])

# are wrong. We have to create seperate coloumns for countries using
# OneHotEncoder

onehotencoder_X = OneHotEncoder(categorical_features = [0])
X = onehotencoder_X.fit_transform(X).toarray()

#*** OnHotEncoder() can not convert from String to Float. So first convert to integer
# using LabelEncoder() ...



# --------------- Splitting into Training & Test Set -----------------

from sklearn.cross_validation import train_test_split

X_train, X_test, Y_train, Y_Test = train_test_split(X, Y, # array i.e the matrix
                                                    test_size = 0.2, # 20% test data
                                                    random_state = 0) # 


# --------------- Feature Scaling ------------

# Feature Scaling = to transform all the variable to same scale

from sklearn.preprocessing import StandardScaler

sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train) # we have to to fit the sc scaler and then transform
X_test = sc_X.transform(X_test) # as we have made the fit to training set, in test set
                                # only transform is required

# is scalling is required on Dummy variable ? It depends on situations
# for Y scaling is not required                               

File: 1-Data Preprocessing/R/Feature_Scaling/feature_scaling.R

#First you have to set the working directory from "More" option menu

# --------- Importing the dataset -------------
dataset = read.csv('Data.csv')

# --------- Taking care of missing data individual coloumn ----------
dataset$Age = ifelse(is.na(dataset$Age), # condition checking:: if value in cloumn Age is missing or not
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)), # if True :: then do the Mean
                     dataset$Age) #if False :: then print Age

dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Salary)

# -------- Encoding categorical data ------

#factor function = convert catagorical variables to numeric values.
#                  it will see the variables as factors.
#                  we can change the catagory
#                  No need to make three coloumns

dataset$Country = factor(dataset$Country, # data which we want to transform into factors
                         levels = c('France', 'Spain', 'Germany'),#names of catagorical coloumns
                                                                  # c is the vector of levels
                         labels = c(1, 2, 3)) #labels value is to give differentiate but not to compare
                        
dataset$Purchased = factor(dataset$Purchased,
                           levels = c('Yes', 'No'),
                           labels = c(0,1))


# --------------- Splitting into Training & Test Set -----------------

#install.packages('caTools') # we have to download this liabrary for splitting

library(caTools) # to include the liabrary to the script

set.seed(123)

#This will return only True and False
#                   True = observation in training set
#                   False = observation in test set

split = sample.split(dataset$Purchased, # put the dependant variable only
                     SplitRatio = 0.8) # Training set 80%


training_set =  subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


#----------- Feature Scaling -----------------------

#training_set =  scale(training_set) # this will show an error as coloumn 1 and 4 
                                    # are factors not numeric
#test_set =  scale(test_set)

training_set[, 2:3] =  scale(training_set[, 2:3])
test_set[, 2:3] =  scale(test_set[, 2:3])



